luatest/cluster.lua

```diff
@@ -0,0 +1,349 @@
+
+            cfg:reload()
+        end)
+    end)
+end
+
+local methods = {
+    each = cluster_each,
+    size = cluster_size,
+    start = cluster_start,
+    start_instance = cluster_start_instance,
+    stop = cluster_stop,
+    drop = cluster_drop,
+    sync = cluster_sync,
+    reload = cluster_reload,
+}
+
+local cluster_mt = {
+    __index = function(self, k)
+        if methods[k] ~= nil then
+            return methods[k]
+        end
+        if self._server_map[k] ~= nil then
+            return self._server_map[k]
+        end
+        return rawget(self, k)
+    end
+}
+
+--- Create a new Tarantool cluster.
+--
+-- @tab g Group of tests.
+-- @tab config Cluster configuration.
+-- @tab[opt] server_opts Extra options passed to server:new().
+-- @tab[opt] opts Cluster options.
+-- @string[opt] opts.dir Specific directory for the cluster.
+-- @return table
+local function new(g, config, server_opts, opts)
+    assert(type(config) == 'table')
+    assert(config._config == nil, "Please provide cbuilder:new():config()")
+    assert(g._cluster == nil)
+
+    -- Prepare a temporary directory and write a configuration
+    -- file.
+    local dir = opts and opts.dir or treegen.prepare_directory({}, {})
+    local config_file_rel = 'config.yaml'
+    local config_file = treegen.write_file(dir, config_file_rel,
+                                           yaml.encode(config))
+
+    -- Collect names of all the instances defined in the config
+    -- in the alphabetical order.
+    local instance_names = instance_names_from_config(config)
+
+    assert(next(instance_names) ~= nil, 'No instances in the supplied config')
+
+    -- Generate luatest server options.
+    server_opts = fun.chain({
+        config_file = config_file,
+        chdir = dir,
+        net_box_credentials = {
+            user = 'client',
+            password = 'secret',
+        },
+    }, server_opts or {}):tomap()
+
+    -- Create luatest server objects.
+    local servers = {}
+    local server_map = {}
+    for _, name in ipairs(instance_names) do
+        local iserver = server:new(fun.chain(server_opts, {
+            alias = name,
+        }):tomap())
+        table.insert(servers, iserver)
+        server_map[name] = iserver
+    end
+
+    -- Create a cluster object and store it in 'g'.
+    g._cluster = setmetatable({
+        _servers = servers,
+        _server_map = server_map,
+        _dir = dir,
+        _config_file_rel = config_file_rel,
+        _server_opts = server_opts,
+    }, cluster_mt)
+    return g._cluster
+end
+
+-- }}} Replicaset management
+
+-- {{{ Replicaset that can't start
+
+--- Ensure cluster startup error
+--
+-- Starts a all instance of a cluster from the given config and
+-- ensure that all the instances fails to start and reports the
+-- given error message.
+--
+-- @tab g Group of tests.
+-- @tab config Cluster configuration.
+-- @string exp_err Expected error message.
+local function startup_error(g, config, exp_err)
```

Comment:
The same remark regarding eliminating the test group argument.

Reply:
Applied. Thank you.